bits.go

```diff
@@ -266,3 +267,50 @@ func fromDeltas(data []byte, buf []uint32) []uint32 {
 	}
 	return buf
 }
+
+// condenseRuneOffsets converts the mostly-predictable runeOffset input
+// into a shorter form based on tuples of the unexpected values.
+// The input is roughly a sequence of y values [100*i for i in range(n)],
+// so we just store (x, y) points where the expectation is violated.
+//
+// This is very rarely a slight pessimization, on repos where there are frequent
+// non-ASCII characters.
+func condenseRuneOffsets(off []uint32) []uint32 {
+	expected := uint32(0)
+	tmp := []uint32{}
+	for x, y := range off {
+		if y != expected {
+			tmp = append(tmp, uint32(x)*runeOffsetFrequency, y)
+			expected = y
+		}
+		expected += runeOffsetFrequency
+	}
+	out := make([]uint32, len(tmp))
+	copy(out, tmp)
+	return out
+}
+
+func runeOffsetLookup(sample []uint32, absR uint32) (uint32, uint32) {
+	left := absR % runeOffsetFrequency
+	sampleOffset := absR - left
+	slen := len(sample)
+	if slen == 0 {
+		return sampleOffset, left
+	}
+	idx := sort.Search(slen/2, func(i int) bool {
+		return sampleOffset >= sample[slen-1-2*i]
+	})
```

Comment:
There is a lot of handling corner cases here, which makes me think we can add a helper/etc to reduce the amount of complexity.

How about adding a sort.Search wrapper which returns the last index which is true? I think this is a working imlpementation:

```go
{
  // ...
  idx := lastTrue(len(sample), func(i int) bool {
    return sample[i].runeOffset <= sampleOffset
  })
  
  if idx < 0 {
    return sampleOffset, left
  }
  
  return sample[idx].byteOffset + (sampleOffset - sample[idx].runeOffset), left
}

func lastTrue(length int, f func(int) bool) int {
    return sort.Search(length, func(i int) bool { return !f(i) }) - 1
}
```

Reply:
I think it's clear enough with the new code. That lastTrue function won't work-- sort.Search has a different invariant: `Search uses binary search to find and return the smallest index i in [0, n) at which f(i) is true, assuming that on the range [0, n), f(i) == true implies f(i+1) == true` -- that is, the predicate can *only* shift from false to true as i increases, not from true to false.

Reply:
FYI see what I did with `lastTrue`. I inverted `f(i)` so that we still respect `sort.Search`. Will take al ook at the new code :)